Detangle Slice and fix mixed-dimension in-place reductions

mbauman opened the pull request:
Well this was a fun adventure.  Here I thought fixing #28928 would just be a simple conversion to `ntuple`… it turned into a big rabbit hole that glaringly exposed the need to split our two uses of `Base.Slice` between flagging "whole dimensions" for SubArray and offset axes.  This is something @timholy and I had talked about for quite some time and I thought would be more challenging and not a terribly pressing issue, but it turns out this conflation was producing incorrect results in our reductions.

mbauman committed:
Detangle Slice between whole dimensions and axes

We use axes in many downstream computations that may not re-index directly into the original array, so we add a second type parameter to `Slice` that is turned on when converting `:` in indexing expressions -- and really only `SubArray` cares about.

mbauman committed:
Reduce allocations for in-place reductions and fix mixed-dimensionality edge-cases

Alleviates #28928 but does not completely remove allocations due to the allocation of the view that gets passed to `map!`.

andreasnoack commented on test/offsetarray.jl:
```diff
@@ -214,11 +214,11 @@ targets2 = ["(1.0, 1.0)",
-    @test String(take!(io)) == targets1[n+1]
-    show(IOContext(io, :limit => true), MIME("text/plain"), (a,a))
-    @test String(take!(io)) == targets2[n+1]
+    # a = OffsetArray(fill(1.,ntuple(d->1,n)), ntuple(identity,n))
```
Did you forget to uncomment these or is it intentional?

mbauman replied:
Oh, yeah, this is unintentional.  The test code needs to be re-jiggered for the new package/module names — these tests break if you just `include` the test file.  The naive things I tried for 60 seconds didn't work so I just gave up for now and temporarily commented them out.

mbauman commented on stdlib/SparseArrays/src/sparsematrix.jl:
```diff
@@ -42,7 +42,7 @@ size(S::SparseMatrixCSC) = (S.m, S.n)
 const SparseMatrixCSCView{Tv,Ti} =
     SubArray{Tv,2,SparseMatrixCSC{Tv,Ti},
-        Tuple{Base.Slice{Base.OneTo{Int}},I}} where {I<:AbstractUnitRange}
+        Tuple{Base.Slice{Base.OneTo{Int},true},I}} where {I<:AbstractUnitRange}
```
This (and the accompanying change below) is worth pointing out: in nearly all cases, `Base.Slice` continues working as it used to, but here parametric invariance means that this change broke this type-alias into something that is impossible to match.

So this is indeed slightly breaking.

StefanKarpinski replied:
> So this is indeed slightly breaking.

Policy: if we run it on all registered packages and nothing breaks, then slightly breaking is fine in the next minor release, i.e. Julia 1.1.0 (but not in Julia 1.0.1, of course).

timholy replied:
What would it take to run this before merging this PR? I ask because if required there is an alternative: keep `Slice` meaning what it always has and a new type (`IdentityRange`?) which is the equivalent of this PR's `Slice{T,false}`.

mbauman replied:
Yeah, introducing a new type is probably the wiser tack here.  I had been thinking that this would be less breaking, but just introducing the new type and allowing OffsetArrays to gradually begin using it seems like a better plan.

timholy commented on base/reducedim.jl:
```diff
@@ -218,17 +218,20 @@ Extract first entry of slices of array A into existing array R.
     map!(f, R, view(A, t...))
 end
+# We know that the axes of R and A are compatible, but R might have a different number of
+# dimensions than A, which is trickier than it seems due to offset arrays and type stability
```
Nice to have this inferrable/nonallocating.

timholy commented on test/offsetarray.jl:
```diff
@@ -489,3 +489,22 @@ end
+@testset "in-place reductions with mismatched dimensionalities" begin
+    B = OffsetArray(reshape(1:24, 4, 3, 2), -5, 6, -7)
+    for R in (fill(0, -4:-1), fill(0, -4:-1, 7:7), fill(0, -4:-1, 7:7, -6:-6))
+        @test @inferred(maximum!(R, B)) == reshape(maximum(B, dims=(2,3)), axes(R)) == reshape(21:24, axes(R))
```
Given that #28928 is about allocation, should we check `@allocated` also?

timholy replied:
bump this comment (optional, but given the thumbs-up...)

mbauman replied:
Thanks for the reminder!

timholy reviewed (commented):
Beautiful as always. I am really pleased to have the concepts split, :+1: . Most important question is regarding backwards compatibility, how much of an advantage is there in keeping both behaviors under a single "roof" by adding that type parameter? If that's not important then I think this could be made completely non-breaking?

mbauman commented on base/reducedim.jl:
```diff
@@ -218,17 +218,20 @@ Extract first entry of slices of array A into existing array R.
 copyfirst!(R::AbstractArray, A::AbstractArray) = mapfirst!(identity, R, A)
 
-function mapfirst!(f, R::AbstractArray, A::AbstractArray)
+function mapfirst!(f, R::AbstractArray, A::AbstractArray{<:Any,N}) where {N}
```
I had been having trouble getting `ntuple` to specialize with just constant propagation, so I switched to `Val(N)` instead… but you're probably right that `Val(ndims(A))` would do just as well.

mbauman committed:
Introduce a whole new IdentityUnitRange

that we will encourage offset array implementations to use instead of Base.Slice

mbauman commented:
I tried out Tim's suggestion of a completely different type and I like it.  On the downside, it's basically an exact copy of `Slice`'s implementation, but on the upside, it means that the changes to `Slice` are essentially nonbreaking.  

We just introduce a new type (I have named it `IdentityUnitRange`) that's effectively a copy of `Slice` but without the special "entire-dimension" handling.  We can then encourage OffsetArrays.jl to use it instead of `Base.Slice`.  Perhaps we should even export it as a new feature in 1.1.  How do folks feel about the name?  Here's its docstring:

```
   IdentityUnitRange(range::AbstractUnitRange)

 Represent an AbstractUnitRange `range` as an offset vector such that `range[i] == i`.

 `IdentityUnitRange`s are frequently used as axes for offset arrays.
```

mbauman commented:
Let's make this PR happen. It fixes many such issues.  I'm surprised it hasn't picked up any conflicts since I last touched it… let's see if CI is still happy.

timholy commented:
I say merge at will.

mbauman committed:
Test that maximum! allocates less
